rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+
+Both of these solutions for strict schemas would not be suitable for the type of expressiveness we want for type extension, unfortunately.
+
+The behavior we're after is that `unevaluatedProperties` should only validate at the top level of a type. This would allow supertypes to still validate `unevaluatedProperties`, but defer checking if they're used within in subtype.
+
+### Alternative through open/closed schemas
+
+A proposed way to deal with conditional `unevaluatedProperties` is to use custom `$defs` definitions for open and closed variations of entity types.
+**Open by default schema**:
+
+```json
+{
+  "$id": "https://example.com/schema",
+
+  // ... schema contents  ...
+
+  "$defs": {
+    "closed": {
+      "$anchor": "closed",
+      "$ref": "#",
+      "unevaluatedProperties": false
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#closed` results in retrieving a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`.
+
+**Closed by default schema**:
+
+```json
+{
+  "$id": "https://example.com/schema",
+  "$ref": "#open",
+  "unevaluatedProperties": false,
+
+  "$defs": {
+    "schema": {
+      "$anchor": "open"
+      // ... schema contents ...
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#open` results in retrieving an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`.
+
+(Thanks Jason Desrosiers for the suggestions!)
+
+Using the above setup would mean that we need to specify `#open` at the end of type URIs when URIs appear in `allOf`. We would also need to serve type schemas with one of the above structures, such that they conform with JSON Schema. Although this does add some ceremony around extending types, we would end up with a type extension system that would conform to JSON Schema without implicitness or redefining keywords.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
+- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.
+- Duplicating types of "forking" is not solved by this RFC.
```

Comment:
@Alfred-Mountfield what is the current process for follow-up RFCs mentioned here? Should we leave it like this, or create a draft PR and link it?

Reply:
Either or. I think if we **know** it needs a follow-up in the very near future then we should make the draft because it's nice to reference something. But if we aren't 100% going to address it soon then we probably shouldn't leave drafts hanging around.